<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function reference â€¢ chevreulSeurat</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">chevreulSeurat</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.5.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/integration.html">Integration of Seurat objects</a>
    <a class="dropdown-item" href="../articles/pipeline.html">Pipeline</a>
    <a class="dropdown-item" href="../articles/preprocessing.html">Preprocessing</a>
    <a class="dropdown-item" href="../articles/shiny_app.html">Shiny App</a>
    <a class="dropdown-item" href="../articles/visualization.html">Visualization</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/cobriniklab/chevreulSeurat/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="Monocle2_diffex.html">Monocle2_diffex()</a></code> 
        </dt>
        <dd>Monocle2 differential expression</dd>
      </dl><dl><dt>
          
          <code><a href="SetDefaultAssay.html">SetDefaultAssay()</a></code> 
        </dt>
        <dd>Reset default assay</dd>
      </dl><dl><dt>
          
          <code><a href="add_census_slot.html">add_census_slot()</a></code> 
        </dt>
        <dd>add census assay to a seurat object</dd>
      </dl><dl><dt>
          
          <code><a href="add_percent_mito.html">add_percent_mito()</a></code> 
        </dt>
        <dd>Annotate percent mitochondrial reads per cell</dd>
      </dl><dl><dt>
          
          <code><a href="add_read_count_col.html">add_read_count_col()</a></code> 
        </dt>
        <dd>Annotate Low Read Count Category</dd>
      </dl><dl><dt>
          
          <code><a href="allTranscripts.html">allTranscripts()</a></code> 
        </dt>
        <dd>Plot All Transcripts Server</dd>
      </dl><dl><dt>
          
          <code><a href="allTranscriptsui.html">allTranscriptsui()</a></code> 
        </dt>
        <dd>Plot All Transcripts UI Module</dd>
      </dl><dl><dt>
          
          <code><a href="annotate_cell_cycle.html">annotate_cell_cycle()</a></code> 
        </dt>
        <dd>Annotate Cell Cycle</dd>
      </dl><dl><dt>
          
          <code><a href="annotate_excluded.html">annotate_excluded()</a></code> 
        </dt>
        <dd>Annotate Exclusion Criteria</dd>
      </dl><dl><dt>
          
          <code><a href="append_to_project_db.html">append_to_project_db()</a></code> 
        </dt>
        <dd>Update a database of chevreulSeurat projects</dd>
      </dl><dl><dt>
          
          <code><a href="arrange_ptime_heatmaps.html">arrange_ptime_heatmaps()</a></code> 
        </dt>
        <dd>Arrange processed monocle CDS objects containing heatmaps</dd>
      </dl><dl><dt>
          
          <code><a href="assign_clusters_to_cds.html">assign_clusters_to_cds()</a></code> 
        </dt>
        <dd>Assign Clusters to CDS</dd>
      </dl><dl><dt>
          
          <code><a href="build_bigwig_db.html">build_bigwig_db()</a></code> 
        </dt>
        <dd>Create a database of bigwigfiles</dd>
      </dl><dl><dt>
          
          <code><a href="by_row_corr_between_pt_heatmaps.html">by_row_corr_between_pt_heatmaps()</a></code> 
        </dt>
        <dd>calculate row-wise correlation between two dataframes</dd>
      </dl><dl><dt>
          
          <code><a href="calc_cor_across_heatmaps.html">calc_cor_across_heatmaps()</a></code> 
        </dt>
        <dd>calculate correlation between heatmap matrices in processed cds_gene_list</dd>
      </dl><dl><dt>
          
          <code><a href="calc_pseudotime_heatmap.html">calc_pseudotime_heatmap()</a></code> 
        </dt>
        <dd>calculate pseudotime heatmaps</dd>
      </dl><dl><dt>
          
          <code><a href="ccScore.html">ccScore()</a></code> 
        </dt>
        <dd>Cell Cycle Score</dd>
      </dl><dl><dt>
          
          <code><a href="ccScoreui.html">ccScoreui()</a></code> 
        </dt>
        <dd>Cell Cycle Score UI</dd>
      </dl><dl><dt>
          
          <code><a href="cells_selected.html">cells_selected()</a></code> 
        </dt>
        <dd>Title</dd>
      </dl><dl><dt>
          
          <code><a href="celltype_markers.html">celltype_markers</a></code> 
        </dt>
        <dd>Retinal Cell Type Marker Genes</dd>
      </dl><dl><dt>
          
          <code><a href="changeEmbedParams.html">changeEmbedParams()</a></code> 
        </dt>
        <dd>Change Embedding Parameters</dd>
      </dl><dl><dt>
          
          <code><a href="changeEmbedParamsui.html">changeEmbedParamsui()</a></code> 
        </dt>
        <dd>Change Embedding Parameters UI</dd>
      </dl><dl><dt>
          
          <code><a href="chevreulSeuratBox.html">chevreulSeuratBox()</a></code> 
        </dt>
        <dd>custom collapsible box</dd>
      </dl><dl><dt>
          
          <code><a href="clustering_workflow.html">clustering_workflow()</a></code> 
        </dt>
        <dd>Clustering Workflow</dd>
      </dl><dl><dt>
          
          <code><a href="combine_cols.html">combine_cols()</a></code> 
        </dt>
        <dd>Reformat Seurat Object Metadata</dd>
      </dl><dl><dt>
          
          <code><a href="combine_looms.html">combine_looms()</a></code> 
        </dt>
        <dd>Combine Loom Files</dd>
      </dl><dl><dt>
          
          <code><a href="convert_human_seu_to_mouse.html">convert_human_seu_to_mouse()</a></code> 
        </dt>
        <dd>Convert Seurat Objects from Human to Mouse</dd>
      </dl><dl><dt>
          
          <code><a href="convert_monoclev2_to_seuv3.html">convert_monoclev2_to_seuv3()</a></code> 
        </dt>
        <dd>Convert a Seurat V3 object to a Monocle v2 object</dd>
      </dl><dl><dt>
          
          <code><a href="convert_mouse_seu_to_human.html">convert_mouse_seu_to_human()</a></code> 
        </dt>
        <dd>Convert Seurat Objects from Mouse to Human</dd>
      </dl><dl><dt>
          
          <code><a href="convert_seu_list_to_multimodal.html">convert_seu_list_to_multimodal()</a></code> 
        </dt>
        <dd>convert seurat list to multimodal object</dd>
      </dl><dl><dt>
          
          <code><a href="convert_seu_to_cds.html">convert_seu_to_cds()</a></code> 
        </dt>
        <dd>Convert a Seurat Object to a Monocle Cell Data Set</dd>
      </dl><dl><dt>
          
          <code><a href="convert_seuv3_to_monoclev2.html">convert_seuv3_to_monoclev2()</a></code> 
        </dt>
        <dd>Convert a Seurat V3 object to a Monocle v2 object</dd>
      </dl><dl><dt>
          
          <code><a href="convert_symbols_by_species.html">convert_symbols_by_species()</a></code> 
        </dt>
        <dd>Convert gene symbols between mouse and human</dd>
      </dl><dl><dt>
          
          <code><a href="convert_to_h5ad.html">convert_to_h5ad()</a></code> 
        </dt>
        <dd>convert a seurat object to an on-disk anndata object</dd>
      </dl><dl><dt>
          
          <code><a href="convert_v3_to_v5.html">convert_v3_to_v5()</a></code> 
        </dt>
        <dd>Convert seurat object to seurat V5 format</dd>
      </dl><dl><dt>
          
          <code><a href="create_proj_matrix.html">create_proj_matrix()</a></code> 
        </dt>
        <dd>Create a Table of single Cell Projects</dd>
      </dl><dl><dt>
          
          <code><a href="create_project_db.html">create_project_db()</a></code> 
        </dt>
        <dd>Create a database of chevreulSeurat projects</dd>
      </dl><dl><dt>
          
          <code><a href="cross_check_heatmaps.html">cross_check_heatmaps()</a></code> 
        </dt>
        <dd>Title</dd>
      </dl><dl><dt>
          
          <code><a href="cross_species_integrate.html">cross_species_integrate()</a></code> 
        </dt>
        <dd>Integrate Seurat Objects from Mouse to Human</dd>
      </dl><dl><dt>
          
          <code><a href="default_helper.html">default_helper()</a></code> 
        </dt>
        <dd>Default Shiny Helper</dd>
      </dl><dl><dt>
          
          <code><a href="diffex.html">diffex()</a></code> 
        </dt>
        <dd>Differential Expression</dd>
      </dl><dl><dt>
          
          <code><a href="diffexui.html">diffexui()</a></code> 
        </dt>
        <dd>Differential Expression UI</dd>
      </dl><dl><dt>
          
          <code><a href="dot-send_enrichr_query.html">.send_enrichr_query()</a></code> 
        </dt>
        <dd>Gene enrichment using Enrichr.</dd>
      </dl><dl><dt>
          
          <code><a href="dropdownButton.html">dropdownButton()</a></code> 
        </dt>
        <dd>Custom Dropdown Button</dd>
      </dl><dl><dt>
          
          <code><a href="filter_low_rc_cells.html">filter_low_rc_cells()</a></code> 
        </dt>
        <dd>Filter our Cells from Seurat below read count threshold</dd>
      </dl><dl><dt>
          
          <code><a href="filter_merged_seu.html">filter_merged_seu()</a></code> 
        </dt>
        <dd>Filter a Single Seurat Object</dd>
      </dl><dl><dt>
          
          <code><a href="filter_merged_seus.html">filter_merged_seus()</a></code> 
        </dt>
        <dd>Filter a List of Seurat Objects</dd>
      </dl><dl><dt>
          
          <code><a href="filter_rows_to_top.html">filter_rows_to_top()</a></code> 
        </dt>
        <dd>Filter Rows to Top</dd>
      </dl><dl><dt>
          
          <code><a href="findMarkers.html">findMarkers()</a></code> 
        </dt>
        <dd>Find Markers</dd>
      </dl><dl><dt>
          
          <code><a href="findMarkersui.html">findMarkersui()</a></code> 
        </dt>
        <dd>Find Markers UI</dd>
      </dl><dl><dt>
          
          <code><a href="find_all_markers.html">find_all_markers()</a></code> 
        </dt>
        <dd>Find All Markers at a range of resolutions</dd>
      </dl><dl><dt>
          
          <code><a href="flip_pseudotime.html">flip_pseudotime()</a></code> 
        </dt>
        <dd>Flip Pseudotime</dd>
      </dl><dl><dt>
          
          <code><a href="format_new_metadata.html">format_new_metadata()</a></code> 
        </dt>
        <dd>Title</dd>
      </dl><dl><dt>
          
          <code><a href="format_pathway_table.html">format_pathway_table()</a></code> 
        </dt>
        <dd>Title</dd>
      </dl><dl><dt>
          
          <code><a href="gene_test_plot.html">gene_test_plot()</a></code> 
        </dt>
        <dd>Plot Gene expression over pseudotime on reference and query trajectories from cellAlign</dd>
      </dl><dl><dt>
          
          <code><a href="genes_to_transcripts.html">genes_to_transcripts()</a></code> 
        </dt>
        <dd>Gene Symbols to Ensembl Transcript Ids</dd>
      </dl><dl><dt>
          
          <code><a href="getEnrichedPathways.html">getEnrichedPathways()</a></code> 
        </dt>
        <dd>Get enriched pathways based on marker genes from EnrichR.</dd>
      </dl><dl><dt>
          
          <code><a href="get_meta.html">get_meta()</a></code> 
        </dt>
        <dd>Get Metadata for a project</dd>
      </dl><dl><dt>
          
          <code><a href="get_transcripts_from_seu.html">get_transcripts_from_seu()</a></code> 
        </dt>
        <dd>Get Transcripts in Seurat Object</dd>
      </dl><dl><dt>
          
          <code><a href="global_cellalign.html">global_cellalign()</a></code> 
        </dt>
        <dd>Global Alignment of two trajectories with cellAlign</dd>
      </dl><dl><dt>
          
          <code><a href="harmony_integrate.html">harmony_integrate()</a></code> 
        </dt>
        <dd>integrate small datasets with harmony</dd>
      </dl><dl><dt>
          
          <code><a href="human_to_mouse_homologs.html">human_to_mouse_homologs</a></code> 
        </dt>
        <dd>Gene Homologs Between Human and Mouse</dd>
      </dl><dl><dt>
          
          <code><a href="integrateProj.html">integrateProj()</a></code> 
        </dt>
        <dd>Integrate Projects Server Function</dd>
      </dl><dl><dt>
          
          <code><a href="integrateProjui.html">integrateProjui()</a></code> 
        </dt>
        <dd>Integrate Project UI</dd>
      </dl><dl><dt>
          
          <code><a href="integration_workflow.html">integration_workflow()</a></code> 
        </dt>
        <dd>Integration Workflow</dd>
      </dl><dl><dt>
          
          <code><a href="label_transfer.html">label_transfer()</a></code> 
        </dt>
        <dd>Transfer Labels Between Seurat Objects</dd>
      </dl><dl><dt>
          
          <code><a href="learn_graph_by_resolution.html">learn_graph_by_resolution()</a></code> 
        </dt>
        <dd>Learn Monocle Graph by Resolution</dd>
      </dl><dl><dt>
          
          <code><a href="list_plot_types.html">list_plot_types()</a></code> 
        </dt>
        <dd>Collate list of variables to be plotted</dd>
      </dl><dl><dt>
          
          <code><a href="load_bigwigs.html">load_bigwigs()</a></code> 
        </dt>
        <dd>Load Bigwigs</dd>
      </dl><dl><dt>
          
          <code><a href="load_counts_by_tximport.html">load_counts_by_tximport()</a></code> 
        </dt>
        <dd>Run <a href="http://bioconductor.org/packages/release/bioc/html/tximport.html" class="external-link">tximport</a> on a set of cells</dd>
      </dl><dl><dt>
          
          <code><a href="load_meta.html">load_meta()</a></code> 
        </dt>
        <dd>Load Sample Metadata for a given project</dd>
      </dl><dl><dt>
          
          <code><a href="load_seurat_from_proj.html">load_seurat_from_proj()</a></code> 
        </dt>
        <dd>Load Seurat Files from a signle project path</dd>
      </dl><dl><dt>
          
          <code><a href="load_seurat_path.html">load_seurat_path()</a></code> 
        </dt>
        <dd>Read in Gene and Transcript Seurat Objects</dd>
      </dl><dl><dt>
          
          <code><a href="local_cellalign.html">local_cellalign()</a></code> 
        </dt>
        <dd>Align two trajectories via cellalign</dd>
      </dl><dl><dt>
          
          <code><a href="make_bigwig_db.html">make_bigwig_db()</a></code> 
        </dt>
        <dd>Make Bigwig Database</dd>
      </dl><dl><dt>
          
          <code><a href="make_chevreulSeurat_clean_names.html">make_chevreulSeurat_clean_names()</a></code> 
        </dt>
        <dd>Title</dd>
      </dl><dl><dt>
          
          <code><a href="mem_str.html">mem_str()</a></code> 
        </dt>
        <dd>Profile the memory of an s4object</dd>
      </dl><dl><dt>
          
          <code><a href="merge_small_seus.html">merge_small_seus()</a></code> 
        </dt>
        <dd>Merge Small Seurat Objects</dd>
      </dl><dl><dt>
          
          <code><a href="metadata_from_batch.html">metadata_from_batch()</a></code> 
        </dt>
        <dd>Retrieve Metadata from Batch</dd>
      </dl><dl><dt>
          
          <code><a href="minimalSeuratApp.html">minimalSeuratApp()</a></code> 
        </dt>
        <dd>Create a minimal seurat app</dd>
      </dl><dl><dt>
          
          <code><a href="monocle.html">monocle()</a></code> 
        </dt>
        <dd>Monocle Server Module</dd>
      </dl><dl><dt>
          
          <code><a href="monocle_module_heatmap.html">monocle_module_heatmap()</a></code> 
        </dt>
        <dd>Title</dd>
      </dl><dl><dt>
          
          <code><a href="monocleui.html">monocleui()</a></code> 
        </dt>
        <dd>Monocle UI Module</dd>
      </dl><dl><dt>
          
          <code><a href="pad_sample_files.html">pad_sample_files()</a></code> 
        </dt>
        <dd>pad sample numbers to prep for armor</dd>
      </dl><dl><dt>
          
          <code><a href="pathwayEnrichment.html">pathwayEnrichment()</a></code> 
        </dt>
        <dd>pathway enrichment</dd>
      </dl><dl><dt>
          
          <code><a href="pathwayEnrichmentui.html">pathwayEnrichmentui()</a></code> 
        </dt>
        <dd>Title</dd>
      </dl><dl><dt>
          
          <code><a href="plotClustree.html">plotClustree()</a></code> 
        </dt>
        <dd>plot clustree server</dd>
      </dl><dl><dt>
          
          <code><a href="plotClustree_UI.html">plotClustree_UI()</a></code> 
        </dt>
        <dd>plot clustree ui</dd>
      </dl><dl><dt>
          
          <code><a href="plotCoverage.html">plotCoverage()</a></code> 
        </dt>
        <dd>Plot Coverage Module</dd>
      </dl><dl><dt>
          
          <code><a href="plotCoverage_UI.html">plotCoverage_UI()</a></code> 
        </dt>
        <dd>Title</dd>
      </dl><dl><dt>
          
          <code><a href="plotDimRed.html">plotDimRed()</a></code> 
        </dt>
        <dd>Plot Dimensional Reduduction</dd>
      </dl><dl><dt>
          
          <code><a href="plotDimRedui.html">plotDimRedui()</a></code> 
        </dt>
        <dd>Plot Dimensional Reduduction UI</dd>
      </dl><dl><dt>
          
          <code><a href="plotHeatmap.html">plotHeatmap()</a></code> 
        </dt>
        <dd>Plot Heatmap</dd>
      </dl><dl><dt>
          
          <code><a href="plotHeatmapui.html">plotHeatmapui()</a></code> 
        </dt>
        <dd>Plot Heatmap ui</dd>
      </dl><dl><dt>
          
          <code><a href="plotReadCount.html">plotReadCount()</a></code> 
        </dt>
        <dd>Plot Read Count</dd>
      </dl><dl><dt>
          
          <code><a href="plotReadCountui.html">plotReadCountui()</a></code> 
        </dt>
        <dd>Plot Read Count UI</dd>
      </dl><dl><dt>
          
          <code><a href="plotVelocity.html">plotVelocity()</a></code> 
        </dt>
        <dd>RNA Velocity Server Module</dd>
      </dl><dl><dt>
          
          <code><a href="plotVelocityui.html">plotVelocityui()</a></code> 
        </dt>
        <dd>RNA Velocity UI Module</dd>
      </dl><dl><dt>
          
          <code><a href="plotViolin.html">plotViolin()</a></code> 
        </dt>
        <dd>Plot Violin Server</dd>
      </dl><dl><dt>
          
          <code><a href="plotViolinui.html">plotViolinui()</a></code> 
        </dt>
        <dd>Title</dd>
      </dl><dl><dt>
          
          <code><a href="plot_all_ptimes.html">plot_all_ptimes()</a></code> 
        </dt>
        <dd>Plot a Set of Heatmaps Based on a provided pseudotime</dd>
      </dl><dl><dt>
          
          <code><a href="plot_all_transcripts.html">plot_all_transcripts()</a></code> 
        </dt>
        <dd>Plot All Transcripts</dd>
      </dl><dl><dt>
          
          <code><a href="plot_cds.html">plot_cds()</a></code> 
        </dt>
        <dd>Plot a Monocle Cell Data Set</dd>
      </dl><dl><dt>
          
          <code><a href="plot_cell_cycle_distribution.html">plot_cell_cycle_distribution()</a></code> 
        </dt>
        <dd>Plot cell cycle distribution grouped by metadata</dd>
      </dl><dl><dt>
          
          <code><a href="plot_cells.html">plot_cells()</a></code> 
        </dt>
        <dd>Plot cells of a monocle cell data set</dd>
      </dl><dl><dt>
          
          <code><a href="plot_feature.html">plot_feature()</a></code> 
        </dt>
        <dd>Plot Feature</dd>
      </dl><dl><dt>
          
          <code><a href="plot_feature_in_ref_query_ptime.html">plot_feature_in_ref_query_ptime()</a></code> 
        </dt>
        <dd>Plot Expression of a Given Feature of a set of Pseudotimes</dd>
      </dl><dl><dt>
          
          <code><a href="plot_gene_coverage_by_var.html">plot_gene_coverage_by_var()</a></code> 
        </dt>
        <dd>Plot BigWig Coverage for Genes of Interest Colored by a Given Variable</dd>
      </dl><dl><dt>
          
          <code><a href="plot_markers.html">plot_markers()</a></code> 
        </dt>
        <dd>Plot Cluster Marker Genes</dd>
      </dl><dl><dt>
          
          <code><a href="plot_monocle_features.html">plot_monocle_features()</a></code> 
        </dt>
        <dd>Plot feature expression on a Monocle Cell Data Set</dd>
      </dl><dl><dt>
          
          <code><a href="plot_multiple_branches_heatmap.html">plot_multiple_branches_heatmap()</a></code> 
        </dt>
        <dd>Plot monocle pseudotime over multiple branches</dd>
      </dl><dl><dt>
          
          <code><a href="plot_pseudotime.html">plot_pseudotime()</a></code> 
        </dt>
        <dd>Plot pseudotime on a Monocle Cell Data Set</dd>
      </dl><dl><dt>
          
          <code><a href="plot_pseudotime_heatmap.html">plot_pseudotime_heatmap()</a></code> 
        </dt>
        <dd>plot pseudotime heatmap</dd>
      </dl><dl><dt>
          
          <code><a href="plot_readcount.html">plot_readcount()</a></code> 
        </dt>
        <dd>Plot Read Count</dd>
      </dl><dl><dt>
          
          <code><a href="plot_scvelo.html">plot_scvelo()</a></code> 
        </dt>
        <dd>Plot scvelo on embedding plot</dd>
      </dl><dl><dt>
          
          <code><a href="plot_transcript_composition.html">plot_transcript_composition()</a></code> 
        </dt>
        <dd>Plot Transcript Composition</dd>
      </dl><dl><dt>
          
          <code><a href="plot_var.html">plot_var()</a></code> 
        </dt>
        <dd>Plot Metadata Variables</dd>
      </dl><dl><dt>
          
          <code><a href="plot_violin.html">plot_violin()</a></code> 
        </dt>
        <dd>Plot Violin plot</dd>
      </dl><dl><dt>
          
          <code><a href="plotly_settings.html">plotly_settings()</a></code> 
        </dt>
        <dd>Plotly settings</dd>
      </dl><dl><dt>
          
          <code><a href="prep_armor_meta.html">prep_armor_meta()</a></code> 
        </dt>
        <dd>prep armor metadata file</dd>
      </dl><dl><dt>
          
          <code><a href="prep_plot_genes_in_pseudotime.html">prep_plot_genes_in_pseudotime()</a></code> 
        </dt>
        <dd>Title</dd>
      </dl><dl><dt>
          
          <code><a href="prep_scvelo.html">prep_scvelo()</a></code> 
        </dt>
        <dd>scvelo_assay</dd>
      </dl><dl><dt>
          
          <code><a href="prep_slider_values.html">prep_slider_values()</a></code> 
        </dt>
        <dd>Prep Slider Values</dd>
      </dl><dl><dt>
          
          <code><a href="process_monocle_child.html">process_monocle_child()</a></code> 
        </dt>
        <dd>Preprocess a Monocle v2 object for heatmap based on provided pseudotime</dd>
      </dl><dl><dt>
          
          <code><a href="propagate_spreadsheet_changes.html">propagate_spreadsheet_changes()</a></code> 
        </dt>
        <dd>Propagate Metadata Changes</dd>
      </dl><dl><dt>
          
          <code><a href="read_project_db.html">read_project_db()</a></code> 
        </dt>
        <dd>Read a database of chevreulSeurat projects</dd>
      </dl><dl><dt>
          
          <code><a href="record_experiment_data.html">record_experiment_data()</a></code> 
        </dt>
        <dd>Record Experiment Metadata</dd>
      </dl><dl><dt>
          
          <code><a href="reference_integrate.html">reference_integrate()</a></code> 
        </dt>
        <dd>Transfer labels between gene or transcript objects</dd>
      </dl><dl><dt>
          
          <code><a href="reformatMetadataDR.html">reformatMetadataDR()</a></code> 
        </dt>
        <dd>Reformat Seurat Object Metadata Server</dd>
      </dl><dl><dt>
          
          <code><a href="reformatMetadataDRui.html">reformatMetadataDRui()</a></code> 
        </dt>
        <dd>Reformat Seurat Object Metadata UI</dd>
      </dl><dl><dt>
          
          <code><a href="regress_by_features.html">regress_by_features()</a></code> 
        </dt>
        <dd>Regress Seurat Object by Given Set of Genes</dd>
      </dl><dl><dt>
          
          <code><a href="reintegrate_seu.html">reintegrate_seu()</a></code> 
        </dt>
        <dd>Reintegrate (filtered) seurat objects</dd>
      </dl><dl><dt>
          
          <code><a href="rename_from_x_notation.html">rename_from_x_notation()</a></code> 
        </dt>
        <dd>Rename cell ids from annoying old notation</dd>
      </dl><dl><dt>
          
          <code><a href="rename_seurat.html">rename_seurat()</a></code> 
        </dt>
        <dd>Give a new project name to a seurat object</dd>
      </dl><dl><dt>
          
          <code><a href="reorg_seurat_files.html">reorg_seurat_files()</a></code> 
        </dt>
        <dd>Reorganize seurat objects to a multimodal format</dd>
      </dl><dl><dt>
          
          <code><a href="reset_seu_meta.html">reset_seu_meta()</a></code> 
        </dt>
        <dd>Reset Seurat Metadata</dd>
      </dl><dl><dt>
          
          <code><a href="run_BEAM.html">run_BEAM()</a></code> 
        </dt>
        <dd>Run BEAM from Monocle2</dd>
      </dl><dl><dt>
          
          <code><a href="run_census.html">run_census()</a></code> 
        </dt>
        <dd>Run Census</dd>
      </dl><dl><dt>
          
          <code><a href="run_enrichmentbrowser.html">run_enrichmentbrowser()</a></code> 
        </dt>
        <dd>Run Enrichment Browser on Differentially Expressed Genes</dd>
      </dl><dl><dt>
          
          <code><a href="run_hmap.html">run_hmap()</a></code> 
        </dt>
        <dd>Run heatmap from Monocle2</dd>
      </dl><dl><dt>
          
          <code><a href="run_scvelo.html">run_scvelo()</a></code> 
        </dt>
        <dd>scvelo_assay</dd>
      </dl><dl><dt>
          
          <code><a href="run_seurat_de.html">run_seurat_de()</a></code> 
        </dt>
        <dd>Run Seurat Differential Expression</dd>
      </dl><dl><dt>
          
          <code><a href="save_seurat.html">save_seurat()</a></code> 
        </dt>
        <dd>Save seurat object to <project>/output/sce/<feature>_seu.rds</feature></project></dd>
      </dl><dl><dt>
          
          <code><a href="sce_from_tibbles.html">sce_from_tibbles()</a></code> 
        </dt>
        <dd>Create Single Cell Experiment from Tibbles</dd>
      </dl><dl><dt>
          
          <code><a href="set_colnames_txi.html">set_colnames_txi()</a></code> 
        </dt>
        <dd>Set Column Names from <code>tximport</code></dd>
      </dl><dl><dt>
          
          <code><a href="seu_calcn.html">seu_calcn()</a></code> 
        </dt>
        <dd>Calculate Read Count Metrics for a Seurat object
recalculate counts/features per cell for a seurat object</dd>
      </dl><dl><dt>
          
          <code><a href="seu_complex_heatmap.html">seu_complex_heatmap()</a></code> 
        </dt>
        <dd>Plot Annotated Complexheatmap from Seurat object</dd>
      </dl><dl><dt>
          
          <code><a href="seu_from_tibbles.html">seu_from_tibbles()</a></code> 
        </dt>
        <dd>Create a Seurat Object from a set of tibbles</dd>
      </dl><dl><dt>
          
          <code><a href="seu_from_tximport.html">seu_from_tximport()</a></code> 
        </dt>
        <dd>Create a seurat object from output of  <a href="http://bioconductor.org/packages/release/bioc/html/tximport.html" class="external-link">tximport</a> and a table of cell metadata</dd>
      </dl><dl><dt>
          
          <code><a href="seuratApp.html">seuratApp()</a></code> 
        </dt>
        <dd>Create a shiny app for a project on disk</dd>
      </dl><dl><dt>
          
          <code><a href="seurat_cluster.html">seurat_cluster()</a></code> 
        </dt>
        <dd>Run Louvain Clustering at Multiple Resolutions</dd>
      </dl><dl><dt>
          
          <code><a href="seurat_find_markers.html">seurat_find_markers()</a></code> 
        </dt>
        <dd>Find Cell Type Markers in a Seurat Object</dd>
      </dl><dl><dt>
          
          <code><a href="seurat_integrate.html">seurat_integrate()</a></code> 
        </dt>
        <dd>Batch Correct Multiple Seurat Objects</dd>
      </dl><dl><dt>
          
          <code><a href="seurat_integration_pipeline.html">seurat_integration_pipeline()</a></code> 
        </dt>
        <dd>Run Seurat Integration</dd>
      </dl><dl><dt>
          
          <code><a href="seurat_pipeline.html">seurat_pipeline()</a></code> 
        </dt>
        <dd>Run Seurat Pipeline</dd>
      </dl><dl><dt>
          
          <code><a href="seurat_preprocess.html">seurat_preprocess()</a></code> 
        </dt>
        <dd>Preprocess Seurat Object</dd>
      </dl><dl><dt>
          
          <code><a href="seurat_reduce_dimensions.html">seurat_reduce_dimensions()</a></code> 
        </dt>
        <dd>Dimensional Reduction</dd>
      </dl><dl><dt>
          
          <code><a href="stash_marker_features.html">stash_marker_features()</a></code> 
        </dt>
        <dd>Stash Marker Genes in a Seurat Object</dd>
      </dl><dl><dt>
          
          <code><a href="subset_by_meta.html">subset_by_meta()</a></code> 
        </dt>
        <dd>subset by new metadata</dd>
      </dl><dl><dt>
          
          <code><a href="swap_counts_from_feature.html">swap_counts_from_feature()</a></code> 
        </dt>
        <dd>Swap counts from Feature</dd>
      </dl><dl><dt>
          
          <code><a href="tableSelected.html">tableSelected()</a></code> 
        </dt>
        <dd>Create Table of Selected Cells</dd>
      </dl><dl><dt>
          
          <code><a href="tableSelectedui.html">tableSelectedui()</a></code> 
        </dt>
        <dd>Create Table of Selected Cells UI</dd>
      </dl><dl><dt>
          
          <code><a href="techInfo.html">techInfo()</a></code> 
        </dt>
        <dd>Title</dd>
      </dl><dl><dt>
          
          <code><a href="techInfoui.html">techInfoui()</a></code> 
        </dt>
        <dd>Title</dd>
      </dl><dl><dt>
          
          <code><a href="threshold_monocle_genes.html">threshold_monocle_genes()</a></code> 
        </dt>
        <dd>Threshold monocle genes</dd>
      </dl><dl><dt>
          
          <code><a href="transcripts_to_genes.html">transcripts_to_genes()</a></code> 
        </dt>
        <dd>Ensembl Transcript Ids to Gene Symbols</dd>
      </dl><dl><dt>
          
          <code><a href="unite_metadata.html">unite_metadata()</a></code> 
        </dt>
        <dd>Unite metadata</dd>
      </dl><dl><dt>
          
          <code><a href="update_chevreulSeurat_object.html">update_chevreulSeurat_object()</a></code> 
        </dt>
        <dd>Update a chevreulSeurat Object</dd>
      </dl><dl><dt>
          
          <code><a href="update_exp_meta.html">update_exp_meta()</a></code> 
        </dt>
        <dd>Update experiment Metadata</dd>
      </dl><dl><dt>
          
          <code><a href="update_human_gene_symbols.html">update_human_gene_symbols()</a></code> 
        </dt>
        <dd>Update human gene symbols in seurat object</dd>
      </dl><dl><dt>
          
          <code><a href="update_project_db.html">update_project_db()</a></code> 
        </dt>
        <dd>Update a database of chevreulSeurat projects</dd>
      </dl><dl><dt>
          
          <code><a href="update_seu_meta.html">update_seu_meta()</a></code> 
        </dt>
        <dd>Update Seurat Metadata from Project</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Kevin Stachelek, Bhavana Bhat.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

